internal/persistence/postgres/pool.go

```diff
@@ -0,0 +1,154 @@
+	CopyFrom(ctx context.Context, tableName pgx.Identifier, columnNames []string, rowSrc pgx.CopyFromSource) (int64, error)
+	Ping(ctx context.Context) error
+}
+
+// PoolOpenFunc defines a function that can instantiate a Pool instance with the given options.
+type PoolOpenFunc func(opts Opts, logger *zap.Logger) (Pool, error)
+
+type PoolOpts struct {
+	// Optional The name of the postgres pool implementation to use.
+	// By default, only `pgx` is available.
+	Name string
+
+	// MaxConns is the maximum size of the pool. Default to `persistence.DefaultMaxConn`.
+	MaxConns int32
+
+	// MinConns is the minimum size of the pool. Default to `persistence.DefaultMinConn`.
+	MinConns int32
+
+	// MaxConnLifetime is the duration since creation after which a connection will
+	// be automatically closed. Default to `persistence.DefaultMaxConnLifetime`.
+	MaxConnLifetime time.Duration
+
+	// MaxConnIdleTime is the duration after which an idle connection will be automatically
+	// closed by the health check. Default to `persistence.DefaultMaxConnIdleTime`.
+	MaxConnIdleTime time.Duration
+
+	// HealthCheckPeriod is the duration between checks of the health of
+	// idle connections. Default to `persistence.DefaultHealthCheckPeriod`.
+	HealthCheckPeriod time.Duration
+
+	// ReadOnly indicates the pool will be used for read only operations.
+	ReadOnly bool
+}
+
+// Validate ensures the provided Postgres.Pool options are a valid configuration.
+func (opts *PoolOpts) Validate() error {
+	// by validating these values, we avoid panics from pgxpool at connection
+	if opts.MaxConns < 1 {
+		return fmt.Errorf("invalid Pool.MaxConns value: '%d', should be greater than 0", opts.MaxConns)
+	}
+
+	if opts.HealthCheckPeriod < 1 {
+		return fmt.Errorf("invalid Pool.HealthCheckPeriod value: '%d', should be a positive duration", opts.HealthCheckPeriod)
+	}
+
+	return nil
+}
+
+func newPostgresPool(opts Opts, logger *zap.Logger) (Pool, error) {
+	var dbPool Pool
+	var err error
+
+	if poolOpenFunc, present := pools[opts.Pool.Name]; present {
+		dbPool, err = poolOpenFunc(opts, logger)
+		if err != nil {
+			return nil, err
+		}
+		return dbPool, nil
+	}
+
+	return nil, fmt.Errorf("invalid postgres pool '%s'", opts.Pool.Name)
+}
+
+// NewPgxPoolConfig creates and configure a new pgxpool.Config with the given options.
+func NewPgxPoolConfig(opts Opts, logger *zap.Logger) (*pgxpool.Config, error) {
+	dsn, err := opts.DSN(logger)
+	if err != nil {
+		return nil, err
+	}
+
+	config, err := pgxpool.ParseConfig(dsn)
+	if err != nil {
+		return nil, err
+	}
+
+	config.MaxConns = opts.Pool.MaxConns
+	config.MinConns = opts.Pool.MinConns
+	config.MaxConnLifetime = opts.Pool.MaxConnLifetime
+	config.MaxConnIdleTime = opts.Pool.MaxConnIdleTime
+	config.HealthCheckPeriod = opts.Pool.HealthCheckPeriod
+
+	return config, nil
+}
+
+func newPgxPool(opts Opts, logger *zap.Logger) (Pool, error) {
+	logger.Info("using Postgres PGX Pool")
+	ctx := context.Background()
```

Comment:
Let's use a context with a hard-coded timeout of like 15s or so, because we wouldn't want this to hang in the event it can't establish the connection or ping in a decent timeframe.

Reply:
good point, will fix!